snoop/html.py

```diff
@@ -0,0 +1,103 @@
+import re
+from bs4 import BeautifulSoup, UnicodeDammit
+import lxml.html
+import lxml.html.clean
+import lxml.etree
+
+def _extract_links(html):
+    from lxml.etree import fromstring, tostring, HTMLParser
+
+    tree = fromstring(html, parser=HTMLParser())
+
+    url_index = []
+    i = 0
+    for link in tree.findall('.//a'):
+        url = link.get('href')
+        if url:
+            i += 1
+            index_data = {'title': link.text, 'url': url, 'id': i}
+
+            link.tag = 'span'
+            link.text = '{title} [{id}]'.format(**index_data)
+            url_index.append(index_data)
+
+    html = tostring(tree, encoding="UTF-8", method='html').decode('utf-8')
+
+    html += '<br/>\n<br/>\n'
+    for index_data in url_index:
+        html += '[{id}] {title}: \t {url} <br/>\n'.format(**index_data)
+
+    return html
+
+def _create_lxml_html_cleaner():
+    cleaner = lxml.html.clean.Cleaner()
+
+    # Only the tags will be removed, their content will get pulled up into
+    # the parent tag. The idea here is to make it explicit to copy/paste a link
+    # Also remove any tags that would make inline printing bad (html, head, body)
+    cleaner.remove_tags = ['a', 'img', 'head', 'html', 'body']
+
+    # removes all javascript, like onclick=
+    cleaner.javascript = True
+    # removes <script>s
+    cleaner.scripts = True
+    # removes <link>s
+    cleaner.links = True
+    # removes flash, iframes
+    cleaner.embedded = True
+    cleaner.frames = True
+
+    # removes <blink> and <marquee>
+    cleaner.annoying_tags = True
+
+    # may contain exploits for IE6
+    cleaner.comments = True
+    cleaner.processing_instructions = True
+    cleaner.meta = True
+    cleaner.forms = True
+    cleaner.remove_unknown_tags = True
+
+    # remove <style> so they don't propagate through to the rest of the page
+    cleaner.style = True
+    # keep inline styles, as they make the page more readable
+    cleaner.inline_style = False
+
+    return cleaner
+
+LXML_CLEANER = _create_lxml_html_cleaner()
```

Comment:
I think `LXML_CLEANER` should be lowercase. It's not a plain constant value, it's more like a singleton service.